q2_quality_control/_utilities.py

```diff
@@ -0,0 +1,421 @@
+             'Observed / Expected Taxa')):
+        if var:
+            yvals.append(val)
+
+    # at least one metric must be plotted
+    if len(yvals) < 1:
+        raise ValueError("At least one metric must be plotted.")
+
+    return yvals
+
+
+def _evaluate_composition(exp, obs, depth, palette, metadata, plot_tar,
+                          plot_tdr, plot_r_value, plot_r_squared,
+                          plot_observed_features,
+                          plot_observed_features_ratio):
+
+    yvals = _interpret_metric_selection(
+        plot_tar, plot_tdr, plot_r_value, plot_r_squared,
+        plot_observed_features, plot_observed_features_ratio)
+
+    # If metadata are passed, validate and convert to series
+    if metadata is not None:
+        metadata = _load_metadata(metadata)
+        # validate that metadata ids are superset of obs and values (exp ids)
+        # are subset of exp
+        _validate_metadata_is_superset(metadata, obs)
+        _validate_metadata_values_are_subset(metadata, exp)
+        # DROP NANS/ZERO ABUNDANCE FEATURES
+        obs = _drop_nans_zeros(obs)
+        exp = _drop_nans_zeros(exp)
+
+    # if no metadata are passed, we assume that sample IDs correspond directly
+    # between obs and exp tables
+    else:
+        # DROP MISSING SAMPLES/NANS/ZERO ABUNDANCE FEATURES
+        exp, obs = _match_samples_by_index(exp, obs)
+        obs = _drop_nans_zeros(obs)
+        exp = _drop_nans_zeros(exp)
+
+    # TAR/TDR for obs vs. exp at each level
+    results, vectors = _compute_per_level_accuracy(exp, obs, metadata, depth)
+
+    # regplot of taxa at top level
+    composition_regression = _regplot_from_dict(vectors, palette=palette)
+
+    # pointplot on obs vs. exp at maximum level (coming off of the prior loop)
+    score_plot = _pointplot_multiple_y(
+        results, xval='level', yvals=yvals, palette=palette)
+
+    # list taxa that are obs but not exp
+    obs_collapsed = _collapse_table(obs, depth)
+    exp_collapsed = _collapse_table(exp, depth)
+    obs_features = set(obs_collapsed.columns)
+    exp_features = set(exp_collapsed.columns)
+    fp_features, fn_features = _identify_incorrect_classifications(
+        obs_features, exp_features)
+
+    misclassifications, underclassifications, mismatches = \
+        _tally_misclassifications(fp_features, exp_features)
+
+    # PLOT depth of mismatch at maximum level as histogram
+    mismatch_histogram = _plot_histogram(mismatches)
+
+    # generate tables of false postive/negative taxa. Sort alphabetically.
+    fn_features = exp_collapsed[list(fn_features)].T.sort_index()
+    misclassifications = obs_collapsed[misclassifications].T.sort_index()
+    underclassifications = obs_collapsed[underclassifications].T.sort_index()
+
+    return (results, fn_features, misclassifications, underclassifications,
+            composition_regression, score_plot, mismatch_histogram)
+
+
+def _match_samples_by_index(df, df2):
```

Comment:
This functionality might be slightly clearer as a [`pd.join`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.join.html) where `how='inner'`. Also, how about `df_a` and `df_b`, or `df1` and `df2`, instead of the lopsided names it is right now?

Reply:
👍 name changes

I am not sure about `pd.join`, though — I do not want to join these frames, just drop any samples that are not shared. So `pd.join` adds a extra step since I would need to do something like:
```
df_joined = pd.join([df_a, df_b], how='inner')
df_a = df_joined[[df_a.columns]]
df_b = df_joined[[df_b.columns]]
```

Is that what you had in mind?

Reply:
Oops, I misread the code the first time --- lets stick with what you have. Can you add a test where the index ids are numeric strings (e.g. `"1", "2", "3"`), just to make sure that the `.loc` behavior doesn't change out from underneath us in the future.

Reply:
👍 added `test_match_samples_by_index`